refactor: useUsersGuildPins hook

FBalint opened the pull request.

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -8,7 +8,89 @@ import {
+    .filter((result) => result.status === "failure")
+    .map((result) => result.error)
+
+  const pinIds: bigint[] = results
```
This type is redundant, it'll have the proper type since you're already casting result as bigint ;)

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -8,7 +8,89 @@ import {
+        })
+      : []
+
+  const tokenURIs = results.map((result, index) => ({
```
This variable name is a bit misleading, `tokenURI` is a base64 encoded metadata, this variable contains that base64 string, the chainId, and the tokenId too, so let's find a better name for it please

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -8,7 +8,89 @@ import {
+  return { tokenURIs, errors }
+}
+
+const tokenURItoMetadataJSON = (tokenURI: {
```
This is also misleading for the same reason, but we could call it `getMetadataJSON` or something like that, lmk if you have a better name in your mind

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -19,70 +101,48 @@ const fetchGuildPinsOnChain = async (
-
-    if (newTokenId) usersGuildPinIdsOnChain.push(newTokenId)
+  } catch (e) {
+    return { usersPinsMetadataJSONs: [], errors: [{ error: e, status: "failure" }] }
```
Please type this (even just by adding type casts), so we get a nice clean return type for the `fetchGuildPinsOnChain` function!

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -91,20 +151,40 @@ const fetchGuildPins = async ([_, addresses]: [string, User["addresses"]]) => {
   )
 
-  return responseArray.flat()
+  const { allUsersPins, allErrors } = responseArray.reduce(
```
This reduce unfortunately returns `any[]`s, so let's add a type to it ;)

BrickheadJohnny reviewed (changes requested):
Great refactor, thank you! Added a couple of comments, I think we could make this code a little bit easier to understand with some minor changes 👍

FBalint commented:
> Great refactor, thank you! Added a couple of comments, I think we could make this code a little bit easier to understand with some minor changes 👍

Thanks! I updated the typing, the naming, and fixed an error handling issue. Let me know what you think!

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -8,7 +8,101 @@ import {
+type TokenWithMetadata = {
+  chainId: Chains
+  tokenId: number
+  image: string
```
I think image is already defined in `GuildPinMetadata` ;)

FBalint replied:
Nice, do you have a tip for noticing these typing opportunities?

BrickheadJohnny replied:
I'll be honest, I noticed by chance :D
But, I guess you'll just get to know the codebase better over time and you'll notice these too ;)

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -91,20 +181,40 @@ const fetchGuildPins = async ([_, addresses]: [string, User["addresses"]]) => {
-    ({ chainId: number; tokenId: number } & GuildPinMetadata)[]
-  >(shouldFetch ? ["guildPins", evmAddresses] : null, fetchGuildPins)
+  const swrData = useSWRImmutable(
+    shouldFetch ? ["guildPins", addresses, includeTestnets] : null,
```
We should pass `evmAddresses` here

BrickheadJohnny commented on src/hooks/useUsersGuildPins.ts:
```diff
@@ -91,20 +181,40 @@ const fetchGuildPins = async ([_, addresses]: [string, User["addresses"]]) => {
+  return {
+    ...swrData,
+    data: swrData.data?.usersPins,
+    pinFetchErrors: swrData.data?.errors,
```
I think we could just leave this prop out and instead return a boolean in the `error` prop (`swrData.data?.errors?.length > 0`), since we won't use the actual errors anywhere

BrickheadJohnny reviewed (approved):
🚀
